File: app/buildz/tmp/kapt3/stubs/debug/com/bits/bee/bpmc/presentation/ui/pos/addon/AddOnAdapter.java
package com.bits.bee.bpmc.presentation.ui.pos.addon;

import java.lang.System;

/**
 * Created by aldi on 15/08/22.
 */
@kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u0000T\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\b\n\u0002\b\u0003\n\u0002\u0010\u000b\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0010\u0002\n\u0002\b\u0004\n\u0002\u0018\u0002\n\u0002\b\u0005\u0018\u00002\u000e\u0012\u0004\u0012\u00020\u0002\u0012\u0004\u0012\u00020\u00030\u0001:\u0003\u001e\u001f Bq\u0012\b\b\u0002\u0010\u0004\u001a\u00020\u0005\u0012\b\b\u0002\u0010\u0006\u001a\u00020\u0005\u0012\b\b\u0002\u0010\u0007\u001a\u00020\u0005\u0012\b\b\u0002\u0010\b\u001a\u00020\t\u0012\b\b\u0002\u0010\n\u001a\u00020\t\u0012\u0006\u0010\u000b\u001a\u00020\f\u0012\n\b\u0002\u0010\r\u001a\u0004\u0018\u00010\u000e\u0012\u0006\u0010\u000f\u001a\u00020\u0010\u0012\u001c\u0010\u0011\u001a\u0018\u0012\u0014\u0012\u0012\u0012\u0004\u0012\u00020\u00020\u0013j\b\u0012\u0004\u0012\u00020\u0002`\u00140\u0012\u00a2\u0006\u0002\u0010\u0015J\u0018\u0010\u0016\u001a\u00020\u00172\u0006\u0010\u0018\u001a\u00020\u00032\u0006\u0010\u0019\u001a\u00020\u0005H\u0016J\u0018\u0010\u001a\u001a\u00020\u00032\u0006\u0010\u001b\u001a\u00020\u001c2\u0006\u0010\u001d\u001a\u00020\u0005H\u0016R\u000e\u0010\u000b\u001a\u00020\fX\u0082\u0004\u00a2\u0006\u0002\n\u0000R$\u0010\u0011\u001a\u0018\u0012\u0014\u0012\u0012\u0012\u0004\u0012\u00020\u00020\u0013j\b\u0012\u0004\u0012\u00020\u0002`\u00140\u0012X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\n\u001a\u00020\tX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\b\u001a\u00020\tX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u000f\u001a\u00020\u0010X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0006\u001a\u00020\u0005X\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u0010\u0010\r\u001a\u0004\u0018\u00010\u000eX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0007\u001a\u00020\u0005X\u0082\u000e\u00a2\u0006\u0002\n\u0000\u00a8\u0006!"}, d2 = {"Lcom/bits/bee/bpmc/presentation/ui/pos/addon/AddOnAdapter;", "Landroidx/recyclerview/widget/ListAdapter;", "Lcom/bits/bee/bpmc/domain/model/Item;", "Landroidx/recyclerview/widget/RecyclerView$ViewHolder;", "lastPosition", "", "selectedPosition", "selectionPosition", "isMultiSelect", "", "isMultiQty", "addOnI", "Lcom/bits/bee/bpmc/presentation/ui/pos/addon/AddOnAdapter$AddOnI;", "selectionAdapter", "Lcom/bits/bee/bpmc/presentation/ui/pos/addon/SelectionAdapter;", "lifecycleScope", "Landroidx/lifecycle/LifecycleOwner;", "addOnSelected", "Landroidx/lifecycle/LiveData;", "Ljava/util/ArrayList;", "Lkotlin/collections/ArrayList;", "(IIIZZLcom/bits/bee/bpmc/presentation/ui/pos/addon/AddOnAdapter$AddOnI;Lcom/bits/bee/bpmc/presentation/ui/pos/addon/SelectionAdapter;Landroidx/lifecycle/LifecycleOwner;Landroidx/lifecycle/LiveData;)V", "onBindViewHolder", "", "holder", "position", "onCreateViewHolder", "parent", "Landroid/view/ViewGroup;", "viewType", "AddOnI", "DiffCallback", "ViewHolder", "app_debug"})
public final class AddOnAdapter extends androidx.recyclerview.widget.ListAdapter<com.bits.bee.bpmc.domain.model.Item, androidx.recyclerview.widget.RecyclerView.ViewHolder> {
    private int lastPosition;
    private int selectedPosition;
    private int selectionPosition;
    private final boolean isMultiSelect = false;
    private final boolean isMultiQty = false;
    private final com.bits.bee.bpmc.presentation.ui.pos.addon.AddOnAdapter.AddOnI addOnI = null;
    private final com.bits.bee.bpmc.presentation.ui.pos.addon.SelectionAdapter selectionAdapter = null;
    private final androidx.lifecycle.LifecycleOwner lifecycleScope = null;
    private final androidx.lifecycle.LiveData<java.util.ArrayList<com.bits.bee.bpmc.domain.model.Item>> addOnSelected = null;
    
    public AddOnAdapter(int lastPosition, int selectedPosition, int selectionPosition, boolean isMultiSelect, boolean isMultiQty, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.presentation.ui.pos.addon.AddOnAdapter.AddOnI addOnI, @org.jetbrains.annotations.Nullable()
    com.bits.bee.bpmc.presentation.ui.pos.addon.SelectionAdapter selectionAdapter, @org.jetbrains.annotations.NotNull()
    androidx.lifecycle.LifecycleOwner lifecycleScope, @org.jetbrains.annotations.NotNull()
    androidx.lifecycle.LiveData<java.util.ArrayList<com.bits.bee.bpmc.domain.model.Item>> addOnSelected) {
        super(null);
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public androidx.recyclerview.widget.RecyclerView.ViewHolder onCreateViewHolder(@org.jetbrains.annotations.NotNull()
    android.view.ViewGroup parent, int viewType) {
        return null;
    }
    
    @java.lang.Override()
    public void onBindViewHolder(@org.jetbrains.annotations.NotNull()
    androidx.recyclerview.widget.RecyclerView.ViewHolder holder, int position) {
    }
    
    @kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u0000.\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0004\n\u0002\u0010\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0004\n\u0002\u0018\u0002\n\u0002\b\u0005\n\u0002\u0010\b\n\u0000\b\u0086\u0004\u0018\u00002\u00020\u0001B\r\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u00a2\u0006\u0002\u0010\u0004J\u0010\u0010\u0007\u001a\u00020\b2\u0006\u0010\t\u001a\u00020\nH\u0002J\u000e\u0010\u000b\u001a\u00020\b2\u0006\u0010\f\u001a\u00020\nJ\u0010\u0010\r\u001a\u00020\b2\u0006\u0010\u000e\u001a\u00020\u000fH\u0002J\u0010\u0010\u0010\u001a\u00020\b2\u0006\u0010\t\u001a\u00020\nH\u0002J\u0010\u0010\u0011\u001a\u00020\b2\u0006\u0010\t\u001a\u00020\nH\u0002J\u0010\u0010\u0012\u001a\u00020\b2\u0006\u0010\t\u001a\u00020\nH\u0002J\u0018\u0010\u0013\u001a\u00020\b2\u0006\u0010\u0014\u001a\u00020\u00152\u0006\u0010\t\u001a\u00020\nH\u0002R\u0011\u0010\u0002\u001a\u00020\u0003\u00a2\u0006\b\n\u0000\u001a\u0004\b\u0005\u0010\u0006\u00a8\u0006\u0016"}, d2 = {"Lcom/bits/bee/bpmc/presentation/ui/pos/addon/AddOnAdapter$ViewHolder;", "Landroidx/recyclerview/widget/RecyclerView$ViewHolder;", "binding", "Lcom/bits/bee/bpmc/databinding/ItemAddonBinding;", "(Lcom/bits/bee/bpmc/presentation/ui/pos/addon/AddOnAdapter;Lcom/bits/bee/bpmc/databinding/ItemAddonBinding;)V", "getBinding", "()Lcom/bits/bee/bpmc/databinding/ItemAddonBinding;", "addItem", "", "item", "Lcom/bits/bee/bpmc/domain/model/Item;", "bind", "model", "bindView", "qty", "Ljava/math/BigDecimal;", "deleteItem", "doClick", "minItem", "setSelected", "position", "", "app_debug"})
    public final class ViewHolder extends androidx.recyclerview.widget.RecyclerView.ViewHolder {
        @org.jetbrains.annotations.NotNull()
        private final com.bits.bee.bpmc.databinding.ItemAddonBinding binding = null;
        
        public ViewHolder(@org.jetbrains.annotations.NotNull()
        com.bits.bee.bpmc.databinding.ItemAddonBinding binding) {
            super(null);
        }
        
        @org.jetbrains.annotations.NotNull()
        public final com.bits.bee.bpmc.databinding.ItemAddonBinding getBinding() {
            return null;
        }
        
        public final void bind(@org.jetbrains.annotations.NotNull()
        com.bits.bee.bpmc.domain.model.Item model) {
        }
        
        private final void bindView(java.math.BigDecimal qty) {
        }
        
        private final void doClick(com.bits.bee.bpmc.domain.model.Item item) {
        }
        
        private final void setSelected(int position, com.bits.bee.bpmc.domain.model.Item item) {
        }
        
        private final void addItem(com.bits.bee.bpmc.domain.model.Item item) {
        }
        
        private final void minItem(com.bits.bee.bpmc.domain.model.Item item) {
        }
        
        private final void deleteItem(com.bits.bee.bpmc.domain.model.Item item) {
        }
    }
    
    @kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u0000\u0018\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0010\u000b\n\u0002\b\u0004\u0018\u00002\b\u0012\u0004\u0012\u00020\u00020\u0001B\u0005\u00a2\u0006\u0002\u0010\u0003J\u0018\u0010\u0004\u001a\u00020\u00052\u0006\u0010\u0006\u001a\u00020\u00022\u0006\u0010\u0007\u001a\u00020\u0002H\u0016J\u0018\u0010\b\u001a\u00020\u00052\u0006\u0010\u0006\u001a\u00020\u00022\u0006\u0010\u0007\u001a\u00020\u0002H\u0016\u00a8\u0006\t"}, d2 = {"Lcom/bits/bee/bpmc/presentation/ui/pos/addon/AddOnAdapter$DiffCallback;", "Landroidx/recyclerview/widget/DiffUtil$ItemCallback;", "Lcom/bits/bee/bpmc/domain/model/Item;", "()V", "areContentsTheSame", "", "oldItem", "newItem", "areItemsTheSame", "app_debug"})
    public static final class DiffCallback extends androidx.recyclerview.widget.DiffUtil.ItemCallback<com.bits.bee.bpmc.domain.model.Item> {
        
        public DiffCallback() {
            super();
        }
        
        @java.lang.Override()
        public boolean areItemsTheSame(@org.jetbrains.annotations.NotNull()
        com.bits.bee.bpmc.domain.model.Item oldItem, @org.jetbrains.annotations.NotNull()
        com.bits.bee.bpmc.domain.model.Item newItem) {
            return false;
        }
        
        @java.lang.Override()
        public boolean areContentsTheSame(@org.jetbrains.annotations.NotNull()
        com.bits.bee.bpmc.domain.model.Item oldItem, @org.jetbrains.annotations.NotNull()
        com.bits.bee.bpmc.domain.model.Item newItem) {
            return false;
        }
    }
    
    @kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u0000\u0018\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0000\n\u0002\u0010\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0003\bf\u0018\u00002\u00020\u0001J\u0010\u0010\u0002\u001a\u00020\u00032\u0006\u0010\u0004\u001a\u00020\u0005H&J\u0010\u0010\u0006\u001a\u00020\u00032\u0006\u0010\u0004\u001a\u00020\u0005H&J\u0010\u0010\u0007\u001a\u00020\u00032\u0006\u0010\u0004\u001a\u00020\u0005H&\u00a8\u0006\b"}, d2 = {"Lcom/bits/bee/bpmc/presentation/ui/pos/addon/AddOnAdapter$AddOnI;", "", "addOrRemoveItem", "", "item", "Lcom/bits/bee/bpmc/domain/model/Item;", "onDeselect", "onItemClick", "app_debug"})
    public static abstract interface AddOnI {
        
        public abstract void onItemClick(@org.jetbrains.annotations.NotNull()
        com.bits.bee.bpmc.domain.model.Item item);
        
        public abstract void onDeselect(@org.jetbrains.annotations.NotNull()
        com.bits.bee.bpmc.domain.model.Item item);
        
        public abstract void addOrRemoveItem(@org.jetbrains.annotations.NotNull()
        com.bits.bee.bpmc.domain.model.Item item);
    }
}

File: app/src/main/java/com/bits/bee/bpmc/presentation/ui/pos/addon/AddOnAdapter.kt
package com.bits.bee.bpmc.presentation.ui.pos.addon

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.constraintlayout.widget.ConstraintSet
import androidx.core.view.isVisible
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LiveData
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bits.bee.bpmc.R
import com.bits.bee.bpmc.databinding.ItemAddonBinding
import com.bits.bee.bpmc.domain.model.Item
import com.bits.bee.bpmc.utils.CurrencyUtils
import java.math.BigDecimal

/**
 * Created by aldi on 15/08/22.
 */
class AddOnAdapter(
    private var lastPosition : Int = -1,
    private var selectedPosition : Int = -1,
    private var selectionPosition : Int = 0,
    private val isMultiSelect : Boolean = false,
    private val isMultiQty : Boolean = false,
    private val addOnI : AddOnI,
    private val lifecycleScope : LifecycleOwner,
    private val addOnSelected : LiveData<ArrayList<Item>>,
) : ListAdapter<Item, RecyclerView.ViewHolder>(DiffCallback()) {

    override fun submitList(list: MutableList<Item>?) {
        if(!isMultiSelect){
            list!!.forEachIndexed{ index, item ->
                if(item.qty > BigDecimal.ZERO) {
                    selectedPosition = index
                    lastPosition = -1
                }
            }
        }
        super.submitList(list)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return ViewHolder(
            ItemAddonBinding.inflate(inflater, parent, false)
        )
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val viewHolder = holder as ViewHolder
        viewHolder.bind(getItem(position))
    }

    inner class ViewHolder(val binding : ItemAddonBinding) : RecyclerView.ViewHolder(binding.root){

        fun bind(model : Item){
            binding.apply {
                val qty = model.qty

                root.setOnClickListener {
                    doClick(model)
                }
                rbItem.setOnClickListener {
                    doClick(model)
                }
                cbItem.setOnClickListener {
                    if(cbItem.isChecked)
                        addItem(model)
                    else
                        deleteItem(model)
                }
                ivPlus.setOnClickListener {
                    addItem(model)
                }
                ivMinus.setOnClickListener {
                    minItem(model)
                }

                val context = binding.root.context
                tvItem.text = model.name1
                tvQty.text = CurrencyUtils.formatCurrency(qty)

                tvPrice.isVisible = model.price > BigDecimal.ZERO
                tvPrice.text = context.getString(R.string.mata_uang_nominal, model.crcSymbol, CurrencyUtils.formatCurrency(model.price))

                rbItem.isVisible = !isMultiSelect
                cbItem.isVisible = isMultiSelect

                setSelected(absoluteAdapterPosition, model)

                bindView(model.qty)

                addOnSelected.observe(lifecycleScope) { items ->
                    if(items.isNotEmpty()){
                        val itemIdList = items.map { it.id }
                        if(itemIdList.contains(model.id)){
                            items.lastOrNull{ it.id == model.id }?.let {
                                cbItem.isChecked = it.qty > BigDecimal.ZERO
                                rbItem.isSelected = it.qty > BigDecimal.ZERO
                                model.qty = it.qty
                                tvQty.text = CurrencyUtils.formatCurrency(it.qty)
                                bindView(model.qty)
                            }
                        } else {
                            cbItem.isChecked = false
                            rbItem.isSelected = false

                            bindView(BigDecimal.ZERO)
                        }
                    } else {
                        cbItem.isChecked = false
                        rbItem.isSelected = false

                        bindView(BigDecimal.ZERO)
                    }
                }
            }
        }

        private fun bindView(qty : BigDecimal){
            binding.apply {
                cbItem.isChecked = qty > BigDecimal.ZERO
                rbItem.isSelected = qty > BigDecimal.ZERO

                if(isMultiQty){
                    llQty.isVisible = qty > BigDecimal.ZERO
                } else {
                    llQty.isVisible = false
                    val constraintLayout: ConstraintLayout = clContent
                    val constraintSet = ConstraintSet()
                    constraintSet.clone(constraintLayout)
                    constraintSet.clear(R.id.tvPrice, ConstraintSet.START)
                    constraintSet.clear(R.id.tvPrice, ConstraintSet.TOP)
                    constraintSet.connect(R.id.tvPrice, ConstraintSet.RIGHT, constraintLayout.id, ConstraintSet.RIGHT, 10)
                    constraintSet.connect(R.id.tvPrice, ConstraintSet.TOP, constraintLayout.id, ConstraintSet.TOP)
                    constraintSet.clear(R.id.tvItem, ConstraintSet.BOTTOM)
                    constraintSet.setMargin(R.id.tvItem, ConstraintSet.BOTTOM, 0)
                    constraintSet.setMargin(R.id.tvItem, ConstraintSet.TOP, 0)
                    constraintSet.setMargin(R.id.tvPrice, ConstraintSet.TOP, 0)
                    constraintSet.setMargin(R.id.tvPrice, ConstraintSet.BOTTOM, 0)
                    constraintSet.connect(R.id.tvItem, ConstraintSet.BOTTOM, R.id.view, ConstraintSet.TOP)
                    constraintSet.applyTo(constraintLayout)
                }
            }
        }

        private fun doClick(item: Item) {
            binding.apply {
                if (rbItem.isVisible) {
                    selectedPosition = absoluteAdapterPosition
                    if(item.isVariant && !item.isAddOn)
                        addOnI.onItemClick(item)
                    else {
                        if (!rbItem.isSelected) {
                            item.qty = BigDecimal.ONE
                            addItem(item)
                        } else {
                            deleteItem(item)
                        }
                    }
                    notifyDataSetChanged()
                } else if (cbItem.isVisible) {
                    cbItem.isChecked = !cbItem.isChecked
                    if (cbItem.isChecked) {
                        addItem(item)
                    } else {
                        deleteItem(item)
                    }
                }
            }
        }

        private fun setSelected(position: Int, item: Item) {
            binding.apply {
                if (selectedPosition == position) {
                    if (lastPosition != position) {
                        rbItem.isChecked = true
                        item.qty = BigDecimal.ONE
                        lastPosition = position
                    } else {
                        item.qty = BigDecimal.ZERO
                        rbItem.isChecked = false
                        lastPosition = -1
                        addOnI.onDeselect(item)
                    }
                } else {
                    item.qty = BigDecimal.ZERO
                    rbItem.isChecked = false
                    if(!isMultiSelect)
                        addOnI.onMin(item)
                }
            }
        }

        private fun addItem(item: Item) {
            item.qty = item.qty + BigDecimal.ONE
            addOnI.addOrRemoveItem(item)
        }


        private fun minItem( item: Item) {
            if (item.qty > BigDecimal.ONE) {
                item.qty = item.qty - BigDecimal.ONE
            } else {
                item.qty = BigDecimal.ZERO
            }
            addOnI.addOrRemoveItem(item)
        }

        private fun deleteItem(item: Item) {
            item.qty = BigDecimal.ZERO
            addOnI.addOrRemoveItem(item)
        }
    }


    class DiffCallback : DiffUtil.ItemCallback<Item>() {
        override fun areItemsTheSame(oldItem: Item, newItem: Item): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: Item, newItem: Item): Boolean {
            return oldItem == newItem
        }
    }

    interface AddOnI {
        fun onItemClick(item: Item)
        fun onMin(item : Item)
        fun onDeselect(item: Item)
        fun addOrRemoveItem(item : Item)
    }

}